File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        if (heights.size() == 1) {
            return heights[0];
        }

        vector<int> extended(heights.size() + 2, 0);

        for (int i = 0; i < heights.size(); i++) {
            extended[i + 1] = heights[i];
        }

        deque<int> myQueue;

        myQueue.push_back(0);

        int res = 0;
        for (int i = 1; i < extended.size(); i++) {
            if (extended[i] > extended[myQueue.back()]) {
                myQueue.push_back(i);
            } else if (extended[i] == extended[myQueue.back()]) {
                myQueue.pop_back();
                myQueue.push_back(i);
            } else {
                while (extended[myQueue.back()] > extended[i]) {
                    int mid = myQueue.back();
                    myQueue.pop_back();
                    int left = myQueue.back();
                    int h = extended[mid];
                    int w = i - left - 1;
                    int v = h * w;
                    res = max(res, v);
                }
                myQueue.push_back(i);
            }
        }

        return res;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights = [0] + heights + [0]
        print(heights)
        stack = [0]
        res = 0

        for i in range(1, len(heights)):
            if heights[i] >= heights[stack[-1]]:
                stack.append(i)
            else:
                while stack and heights[stack[-1]] > heights[i]:
                    idx = stack.pop()
                    h = heights[idx]
                    res = max(res, (i - stack[-1] - 1) * h)

                stack.append(i)

        return res